0.12-beta1 - Runtime panic calculating diff

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.12.0-beta1
+ provider.aws v1.60.0-dev20190216h00-dev
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
locals {
	tcp_ports = {
		80 = 0
	}

	udp_ports = {
		53 = 53
	}

	environment_variables = {
		DEBUG = "*"
		NODE_ENV = "dev"
	}

	secret_variables = {

	}

	base_definition = {
		name = "app"
		image = "nginx:latest"
		memory = 512

		// TCP and UDP port mappings
		portMappings = concat([
			// Loop through TCP ports
			for k, v in local.tcp_ports: {
				hostPort = v == "0" ? null : tonumber(v)
				containerPort = tonumber(k)
				protocol = "tcp"
			}
		], [
			// Loop through UDP ports
			for k, v in local.udp_ports: {
				hostPort = v == "0" ? null : tonumber(v)
				containerPort = tonumber(k)
				protocol = "udp"
			}
		])

		// Environment variables to be mapped
		environment = [
			for k, v in local.environment_variables: {
				name = k
				value = v
			}
		]

		// Secrets to be mapped to env vars
		secrets = [
			for k, v in local.secret_variables: {
				name = k
				valueFrom = v
			}
		]
	}
}

resource "aws_ecs_task_definition" "bug" {
	family = "bug"

	container_definitions = jsonencode([
		local.base_definition
	])
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->
[debug.log.gpg](https://github.com/hashicorp/terraform/files/2982955/debug.log)

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
[crash.log.gpg](https://github.com/hashicorp/terraform/files/2982957/crash.log)

### Expected Behavior
<!--
What should have happened?
-->
After the initial apply, changes made to `tcp_ports`, `udp_ports`, `environment_variables` or `secret_variables` should result in a successful diff in the output JSON that is passed to the container definitions property.

### Actual Behavior
<!--
What actually happened?
-->
A runtime panic occurs when any of the properties (mentioned above) are either modified, appended or removed.

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
1. terraform init
2. terraform apply
3. comment out / add / remove an items from any of the maps (mentioned above)
4. terraform plan/apply
5. crash

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
This is not tied directly to the `aws_ecs_task_definition` resource, we have seen this behaviour in other places where dynamic blocks are used. The example above has been recreated for simplicity.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

Sorry for this crash @m13t, and thanks for reporting it!

The stack trace for this error is as follows:

```
panic: runtime error: index out of range

goroutine 49 [running]:
github.com/hashicorp/terraform/command/format.ctySequenceDiff(0xc00025aa00, 0x2, 0x2, 0xc0004bee40, 0x1, 0x1, 0x1, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:1023 +0xb79
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeValueDiff(0xc000835768, 0x29f9c80, 0xc0001ff6c0, 0x2338280, 0xc0001ff6e0, 0x29f9c80, 0xc0004be180, 0x2338280, 0xc0004be1c0, 0x14, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:751 +0x29f
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeValueDiff(0xc000835768, 0x29f9c00, 0xc0006d81c0, 0x23f08c0, 0xc000796de0, 0x29f9c00, 0xc0006d81e0, 0x23f08c0, 0xc0008182d0, 0x10, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:914 +0x131c
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeValueDiff(0xc000835768, 0x29f9c80, 0xc0001ffda0, 0x2338280, 0xc0001ffdc0, 0x29f9c80, 0xc0004beca0, 0x2338280, 0xc0004becc0, 0xc, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:759 +0x53c
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeValueDiff(0xc000835768, 0x29f9b00, 0xc00003c5b0, 0x2357120, 0xc000ea4380, 0x29f9b00, 0xc00003c5b0, 0x2357120, 0xc000ea4540, 0x8, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:604 +0x4f17
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeAttrDiff(0xc000a75768, 0xc00003a8a0, 0x15, 0xc00097a060, 0x29f9b00, 0xc00003c5b0, 0x2357120, 0xc000ea4380, 0x29f9b00, 0xc00003c5b0, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:255 +0x42d
github.com/hashicorp/terraform/command/format.(*blockBodyDiffPrinter).writeBlockBodyDiff(0xc000835768, 0xc00087ebc0, 0x29f9c00, 0xc0006d80f0, 0x23f08c0, 0xc000b92d80, 0x29f9c00, 0xc0006d8118, 0x23f08c0, 0xc000aefc80, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:191 +0x5d7
github.com/hashicorp/terraform/command/format.ResourceChange(0xc000132f00, 0xc00087ebc0, 0xc00045e880, 0x17, 0xc00037f3d0)
	/opt/gopath/src/github.com/hashicorp/terraform/command/format/diff.go:135 +0x629
github.com/hashicorp/terraform/backend/local.(*Local).renderPlan(0xc0004b28f0, 0xc000211810, 0xc00011caa0)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:258 +0x68a
github.com/hashicorp/terraform/backend/local.(*Local).opApply(0xc0004b28f0, 0x29f8200, 0xc0002cad00, 0x29f8200, 0xc0002cad40, 0xc0002faa80, 0xc0002cacc0)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend_apply.go:111 +0x118b
github.com/hashicorp/terraform/backend/local.(*Local).opApply-fm(0x29f8200, 0xc0002cad00, 0x29f8200, 0xc0002cad40, 0xc0002faa80, 0xc0002cacc0)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:305 +0x66
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc00045ed20, 0xc00045ed30, 0xc00045ed40, 0xc0002faa80, 0xc0004b28f0, 0xc0002cacc0, 0xc00045ed10, 0x29f8200, 0xc0002cad00, 0x29f8200, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:356 +0x135
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/opt/gopath/src/github.com/hashicorp/terraform/backend/local/backend.go:340 +0x362
```

This bug seems to be in the code that prints out diffs for list and tuple values:

https://github.com/hashicorp/terraform/blob/e4dfd09be21ee2f69e7f9bd96b6aab2aed6885d4/command/format/diff.go#L1022-L1023

Given the condition on the previous line, it appears that it's `newI` that is out of range here. Given the context, I expect it's occurring while Terraform is trying to show a structural diff of the JSON string assigned to `container_definitions`.

--- a/command/format/diff.go
+++ b/command/format/diff.go
@@ -1067,7 +1067,7 @@ func ctySequenceDiff(old, new []cty.Value) []*plans.Change {
 	var oldI, newI, lcsI int
 	for oldI < len(old) || newI < len(new) || lcsI < len(lcs) {
 		for oldI < len(old) && (lcsI >= len(lcs) || !old[oldI].RawEquals(lcs[lcsI])) {
-			isObjectDiff := old[oldI].Type().IsObjectType() && new[newI].Type().IsObjectType()
+			isObjectDiff := old[oldI].Type().IsObjectType() && (newI >= len(new) || new[newI].Type().IsObjectType())
 			if isObjectDiff && newI < len(new) {
 				ret = append(ret, &plans.Change{
 					Action: plans.Update,

--- a/command/format/diff_test.go
+++ b/command/format/diff_test.go
@@ -761,7 +761,7 @@ func TestResourceChange_JSON(t *testing.T) {
     }
 `,
 		},
-		"JSON list of objects": {
+		"JSON list of objects - adding item": {
 			Action: plans.Update,
 			Mode:   addrs.ManagedResourceMode,
 			Before: cty.ObjectVal(map[string]cty.Value{
@@ -790,6 +790,41 @@ func TestResourceChange_JSON(t *testing.T) {
                 },
               + {
                   + two = "222"
+                },
+            ]
+        )
+    }
+`,
+		},
+		"JSON list of objects - removing item": {
+			Action: plans.Update,
+			Mode:   addrs.ManagedResourceMode,
+			Before: cty.ObjectVal(map[string]cty.Value{
+				"id":         cty.StringVal("i-02ae66f368e8518a9"),
+				"json_field": cty.StringVal(`[{"one": "111"}, {"two": "222"}]`),
+			}),
+			After: cty.ObjectVal(map[string]cty.Value{
+				"id":         cty.UnknownVal(cty.String),
+				"json_field": cty.StringVal(`[{"one": "111"}]`),
+			}),
+			Schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"id":         {Type: cty.String, Optional: true, Computed: true},
+					"json_field": {Type: cty.String, Optional: true},
+				},
+			},
+			RequiredReplace: cty.NewPathSet(),
+			Tainted:         false,
+			ExpectedOutput: `  # test_instance.example will be updated in-place
+  ~ resource "test_instance" "example" {
+      ~ id         = "i-02ae66f368e8518a9" -> (known after apply)
+      ~ json_field = jsonencode(
+          ~ [
+                {
+                    one = "111"
+                },
+              - {
+                  - two = "222"
                 },
             ]
         )

